src/libraries/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs

```diff
@@ -217,7 +217,7 @@ public void MoveTo(string destDirName)
 
             // Windows will throw if the source file/directory doesn't exist, we preemptively check
             // to make sure our cross platform behavior matches .NET Framework behavior.
-            if (!Exists && !FileSystem.FileExists(FullPath))
+            if (!Exists && !FileSystem.DirectoryExists(FullPath))
```

Comment:
Exists does not know if the directory 'really' exists on the file system. (See e.g. my test via a `Directory.CreateDirectory` in #52293) 

I think the behavior should be the same way it was in .net4.x. That is if the `Exists` flag is false we should still check with the file system directly to see if the directory exists or not. (If the exists flag is incorrectly `true` it will fail on the move call to the file system later).

In the end the code that worked in the .net4.x should still work. E.g.:

```csharp
DirectoryInfo tempDirectoryInfo = new DirectoryInfo("t:\\testdir");
Debug.WriteLine(tempDirectoryInfo.Exists);
Directory.CreateDirectory("T:\\testdir");
Debug.WriteLine(tempDirectoryInfo.Exists);
tempDirectoryInfo.MoveTo("t:\\testdir2");
```

Reply:
> Exists does not know if the directory 'really' exists on the file system. (See e.g. my test via a `Directory.CreateDirectory` in #52293)

In the issue you said "Note that on .net5.0 the second tempDirectoryInfo.Exists returns true" - Where is the truth? :-)

Reply:
`Exists` is correct if you create the directory through the `DirectoryInfo` object (as in the first example given in #53393, at least on .net5). But it does not if the directory was created without using the `DirectoryInfo` instance (like in the example given above).

In my opinion the exception should only be thrown when the file system tells us that the directory does not exist and not depend solely on the possibly incorrect `Exist` property. This would guarantee that no exception is thrown in case the directory exists in the file system, this would also match how .net4 code worked and it seems this was also the original intention of the code and `FileExists` was used in error.

Reply:
Good point! You are right, there seems to be a similar issue in `FileInfo.MoveTo`

This code does not throw an exception on .net4 but does on .net5:

```csharp
// Make sure the file does not exist yet
FileInfo tempFileInfo = new FileInfo("t:\\test.txt");
Debug.WriteLine(tempFileInfo.Exists);
File.Create("T:\\test.txt").Dispose();
Debug.WriteLine(tempFileInfo.Exists);
tempFileInfo.MoveTo("t:\\test2.txt");
```

I think the exists check [here](https://github.com/dotnet/runtime/blob/0390d2dcb8e620ad297029596553b44da1a4b6b5/src/libraries/System.IO.FileSystem/src/System/IO/FileInfo.cs#L156) should do the same kind of check via `FileSystem.FileExists`